File: python/Backtracking/subsets.py
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        """
        Approach:
        - Use backtracking (DFS) to generate all possible subsets.
        - At each index, decide to include or exclude the current element.
        - When the end of the list is reached, add the current subset to the result.

        Time Complexity: O(2^n), where n is the length of nums (number of subsets).
        Space Complexity: O(2^n * n), for the result list (all subsets) and recursion stack.
        """
        res = []
        subset = []
        def dfs(i):
            if i >= len(nums):
                res.append(subset.copy())
                return 
            subset.append(nums[i])
            dfs(i + 1)
            subset.pop()
            dfs(i + 1)
        dfs(0)
        return res

# -------------------
# Test Cases
# -------------------
def test_subsets():
    sol = Solution()
    # Test case 1: Example from Leetcode
    nums1 = [1,2,3]
    result1 = sol.subsets(nums1)
    expected1 = [
        [],[1],[2],[3],[1,2],[1,3],[2,3],[1,2,3]
    ]
    assert sorted([tuple(sorted(x)) for x in result1]) == sorted([tuple(sorted(x)) for x in expected1])
    # Test case 2: Single element
    nums2 = [42]
    assert sol.subsets(nums2) == [[],[42]]
    # Test case 3: Empty list
    nums3 = []
    assert sol.subsets(nums3) == [[]]
    print("All test cases passed!")

if __name__ == "__main__":
    test_subsets()

File: C++/Backtracking/subsets.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Approach:
 * - Use DFS (backtracking) to explore all subsets.
 * - At each index, decide whether to include nums[i] or not.
 * - When we reach the end of nums, we save the current subset.
 *
 * Time Complexity: O(2^n) → There are 2^n subsets.
 * Space Complexity: O(2^n * n) → Total size of result + recursion stack.
 */

class Solution {
public:
    vector<vector<int>> res;      // Stores all subsets
    vector<int> subset;           // Current subset being built

    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(0, nums);
        return res;
    }

    void dfs(int i, vector<int>& nums) {
        // If we've considered all elements, save the current subset
        if (i >= nums.size()) {
            res.push_back(subset);
            return;
        }

        // Include nums[i] in the current subset
        subset.push_back(nums[i]);
        dfs(i + 1, nums);

        // Backtrack: remove nums[i] and explore the path without it
        subset.pop_back();
        dfs(i + 1, nums);
    }
};

/*
-----------------------------------------------
📄 Line-by-Line C++ Syntax Explanation:
-----------------------------------------------

#include <bits/stdc++.h>
    → Includes all standard C++ libraries (like vector, iostream, etc).

using namespace std;
    → Avoids writing std:: everywhere (so we can just write vector instead of std::vector).

class Solution { ... };
    → Defines a class named Solution.

vector<vector<int>> res;
    → A 2D vector to store all subsets.

vector<int> subset;
    → A vector to build the current subset during recursion.

vector<vector<int>> subsets(vector<int>& nums) { ... }
    → Public method to start the DFS and return all subsets.

dfs(0, nums);
    → Start DFS from index 0.

void dfs(int i, vector<int>& nums) { ... }
    → Private helper method that performs the DFS from index i.

if (i >= nums.size()) { ... }
    → If index is beyond the last element, we've formed a subset.

res.push_back(subset);
    → Add the current subset to the result.

subset.push_back(nums[i]);
    → Include nums[i] in the current subset.

dfs(i + 1, nums);
    → Recursively move to the next index.

subset.pop_back();
    → Backtrack by removing nums[i] from the current subset.

-----------------------------------------------
⏱ Time & Space:
- Time: O(2^n) → because each element has two choices: include or exclude.
- Space: O(2^n * n) → result + recursion stack.

-----------------------------------------------
*/

